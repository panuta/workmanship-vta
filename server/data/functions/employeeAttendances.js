
export const getMonthlyEmployeesAttendances = async (attendanceMonth) => {
  // TODO => Get data between 26-25 each month
}
