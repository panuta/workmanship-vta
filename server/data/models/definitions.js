import path from 'path'
import { Sequelize, DataTypes } from 'sequelize'

const sequelize = new Sequelize({ dialect: 'sqlite', storage: path.resolve(__dirname, '../../database.sqlite') })

export const SourceFile = sequelize.define('SourceFile', {
  sourceId: { type: DataTypes.UUID, defaultValue: Sequelize.UUIDV4 },
  originalFilename: { type: DataTypes.STRING },
  monthYear: { type: DataTypes.DATEONLY },
  filePath: { type: DataTypes.STRING },
  uploadedDatetime: { type: DataTypes.DATE },
  isSelected: { type: DataTypes.BOOLEAN, defaultValue: false }
})

export const Shift = sequelize.define('Shift', {
  code: { type: DataTypes.STRING, allowNull: false },
  start: { type: DataTypes.STRING },
  end: { type: DataTypes.STRING },
  break: { type: DataTypes.INTEGER }
})

export const Employee = sequelize.define('Employee', {
  code: { type: DataTypes.STRING, allowNull: false },
  salutation: { type: DataTypes.STRING },
  fullName: { type: DataTypes.STRING, allowNull: false },
  nickName: { type: DataTypes.STRING },
  company: { type: DataTypes.STRING, allowNull: false },
  status: { type: DataTypes.STRING, allowNull: false },
  department: { type: DataTypes.STRING, allowNull: false },
  position: { type: DataTypes.STRING },
  startDate: { type: DataTypes.DATEONLY, allowNull: false },
  terminationDate: { type: DataTypes.DATEONLY }
})

export const EmployeeAttendance = sequelize.define('EmployeeAttendance', {
  code: { type: DataTypes.STRING, allowNull: false, unique: 'compositeIndex' },
  attendanceDate: { type: DataTypes.DATEONLY, allowNull: false, unique: 'compositeIndex' },
  shift: { type: DataTypes.STRING },
  compensation: { type: DataTypes.DECIMAL },
  usedCompensation: { type: DataTypes.DECIMAL },
  forceBreak: { type: DataTypes.DECIMAL },
  returnedForceBreak: { type: DataTypes.DECIMAL },
  minuteLate: { type: DataTypes.DECIMAL },
  minuteEarlyLeave: { type: DataTypes.DECIMAL },
  noShow: { type: DataTypes.DECIMAL },
  diligenceAllowance: { type: DataTypes.DECIMAL },
}, {
  indexes: [{ unique: true, fields: ['code', 'attendanceDate'] }]
})
