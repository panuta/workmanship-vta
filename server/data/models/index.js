// eslint-disable-next-line max-classes-per-file
import { Sequelize, DataTypes, Model } from 'sequelize'

import { config as appConfig } from '../../config'

export class SourceFile extends Model {}
export class Shift extends Model {}
export class Employee extends Model {}
export class EmployeeAttendance extends Model {}

let databaseConnection

export const initDatabase = async () => {
  databaseConnection = new Sequelize({
    dialect: 'sqlite',
    storage: appConfig.databaseStorage,
    logging: false
  })

  // Source File
  SourceFile.init({
    originalFilename: { type: DataTypes.STRING },
    dataSourceDate: { type: DataTypes.DATEONLY },
    filePath: { type: DataTypes.STRING },
    uploadedDatetime: { type: DataTypes.DATE }
  }, {
    sequelize: databaseConnection,
    modelName: 'SourceFile',
    timestamps: appConfig.databaseTimestamps
  })

  await SourceFile.sync({ alter: false })

  // Shift
  Shift.init({
    code: { type: DataTypes.STRING, allowNull: false },
    start: { type: DataTypes.STRING },
    end: { type: DataTypes.STRING },
    break: { type: DataTypes.INTEGER }
  }, {
    sequelize: databaseConnection,
    modelName: 'Shift',
    timestamps: appConfig.databaseTimestamps
  })

  await Shift.sync({ alter: false })

  // Employee
  Employee.init({
    code: { type: DataTypes.STRING, allowNull: false },
    salutation: { type: DataTypes.STRING },
    fullName: { type: DataTypes.STRING, allowNull: false },
    nickName: { type: DataTypes.STRING },
    company: { type: DataTypes.STRING, allowNull: false },
    status: { type: DataTypes.STRING, allowNull: false },
    department: { type: DataTypes.STRING, allowNull: false },
    position: { type: DataTypes.STRING },
    startDate: { type: DataTypes.DATEONLY, allowNull: false },
    terminationDate: { type: DataTypes.DATEONLY }
  }, {
    sequelize: databaseConnection,
    modelName: 'Employee',
    timestamps: appConfig.databaseTimestamps
  })

  await Employee.sync({ alter: false })

  // Employee Attendance
  EmployeeAttendance.init({
    code: { type: DataTypes.STRING, allowNull: false, unique: 'compositeIndex' },
    attendanceDate: { type: DataTypes.DATEONLY, allowNull: false, unique: 'compositeIndex' },
    shift: { type: DataTypes.STRING },
    compensation: { type: DataTypes.DECIMAL },
    minutesLate: { type: DataTypes.DECIMAL },
    minutesEarlyLeave: { type: DataTypes.DECIMAL },
    overtime: { type: DataTypes.DECIMAL },
    notice: { type: DataTypes.INTEGER }
  }, {
    sequelize: databaseConnection,
    modelName: 'EmployeeAttendance',
    indexes: [{ unique: true, fields: ['code', 'attendanceDate'] }],
    timestamps: appConfig.databaseTimestamps
  })

  await EmployeeAttendance.sync({ alter: false })
}

export const dropDatabase = async () => {
  if(databaseConnection !== undefined) {
    await databaseConnection.drop()
  }
}
