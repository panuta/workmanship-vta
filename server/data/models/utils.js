
// TODO : bulkUpdate base on some key

export const parseDuration = (text) => {
  const [hoursText, minutesText] = text.split('.')
  try {
    return parseInt(hoursText, 10) * 60 + parseInt(minutesText, 10)
  } catch (error) {
    return null
  }
}
