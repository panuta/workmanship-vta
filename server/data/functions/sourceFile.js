import { SourceFile } from '../models/definitions'

/**
 * Check if SourceFile for specified date existed
 */
export const hasSourceFileByDate = async (dataSourceDate) => {
  const sourceFile = await SourceFile.findOne({ where: { dataSourceDate } })
  return sourceFile !== null
}

/**
 * Create SourceFile instance
 */
export const insertSourceFile = async (dataSourceDate, uploadedFilePath, originalFilename) => {
  return SourceFile.create({
    originalFilename,
    dataSourceDate,
    filePath: uploadedFilePath,
    uploadedDatetime: new Date(),
  })
}
