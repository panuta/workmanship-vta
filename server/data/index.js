import { findEmployeeAttendances, findEmployees } from './models'

export const getMonthlyEmployeeAttendances = async (attendanceMonth) => {
  const employees = await findEmployees(attendanceMonth)
  const employeeAttendances = await findEmployeeAttendances(attendanceMonth)

  const employeeMapping = employees.reduce((accumulator, employee) => {
    accumulator[employee.code] = {
      employee,
      vacation: 0
    }
    return accumulator
  }, {})

  employeeAttendances.forEach(employeeAttendance => {
    // Vacation
    if(employeeAttendance.shift === 'พักร้อน') {
      employeeMapping[employeeAttendance.code].vacation += 1
    }


  })

  const result = Object.values(employeeMapping).map(employeeData => {
    const employeeDict = employeeData.employee.toJSON()
    // console.log(employeeDict)
    return Object.assign(employeeDict, {
      vacation: employeeData.vacation
    })
  })

  // console.log(result)
  return result
}
