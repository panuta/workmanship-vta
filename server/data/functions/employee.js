import { Op } from 'sequelize'
import { Employee } from '../models'
import { dateToString, monthPeriod } from '../utils'

export const findEmployees = async (attendanceMonth) => {
  if(attendanceMonth !== undefined) {
    /**
     * Active employee
     * ( startDate is before end of month period )
     * AND
     * ( ( terminationDate is null ) OR ( terminationDate is after start of period ) )
     */
    const [startPeriod, endPeriod] = monthPeriod(attendanceMonth.year(), attendanceMonth.month())
    const startPeriodString = dateToString(startPeriod)
    const endPeriodString = dateToString(endPeriod)
    return Employee.findAll({
      where: {
        [Op.and]: [
          { startDate: {
            [Op.lte]: endPeriodString } },
          { terminationDate: {
            [Op.or]: [
              { [Op.is]: null },
              { [Op.gte]: startPeriodString },
            ] } },
        ]
      }
    })
  }
  return Employee.findAll()
}
