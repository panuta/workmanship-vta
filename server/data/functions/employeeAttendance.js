import _ from 'lodash'
import dayjs from 'dayjs'
import { Op } from 'sequelize'
import { EmployeeAttendance } from '../models'
import { findEmployees } from './employee'

export const getEmployeeAttendances = async (attendanceMonth) => {
  const date = dayjs(attendanceMonth)
  const attendanceStart = date.startOf('month').toDate()
  const attendanceEnd = date.endOf('month').toDate()

  return EmployeeAttendance.findAll({
    where: {
      attendanceDate: {
        [Op.between]: [attendanceStart, attendanceEnd]
      }
    }
  })
}

export const getMonthlyEmployeesAttendances = async (attendanceMonth) => {
  const activeEmployees = await findEmployees(attendanceMonth)
  const employeesAttendances = await getEmployeeAttendances(attendanceMonth)

  // Convert list of employees to employee mapping
  const employeeMapping = activeEmployees.reduce((accumulator, employee) => {
    accumulator[employee.code] = {
      employee,
      notice: 0,
      vacation: 0,
      sickLeave: 0,
      casualLeave: 0,
      compensation: 0,
      minutesLate: 0,
      minutesEarlyLeave: 0,
      noShow: 0,
      diligenceAllowance: 0,
    }
    return accumulator
  }, {})

  employeesAttendances.forEach(employeeAttendance => {
    // Only for employees that is valid within this month
    if(_.has(employeeMapping, [employeeAttendance.code])) {
      const increaseWhenShiftIs = (shiftName, valueName) => {
        if(employeeAttendance.shift === shiftName) {
          employeeMapping[employeeAttendance.code][valueName] += 1
        }
      }

      const increaseByValue = (valueName) => {
        if(_.isNumber(employeeAttendance[valueName])) {
          employeeMapping[employeeAttendance.code][valueName] += employeeAttendance[valueName]
        }
      }

      increaseWhenShiftIs('พักร้อน', 'vacation')
      increaseWhenShiftIs('ลาป่วย', 'sickLeave')
      increaseWhenShiftIs('ลากิจ', 'casualLeave')

      increaseByValue('compensation')
      increaseByValue('usedCompensation')
      increaseByValue('forceBreak')
      increaseByValue('returnedForceBreak')
      increaseByValue('minuteLate')
      increaseByValue('minuteEarlyLeave')
      increaseByValue('noShow')
      increaseByValue('diligenceAllowance')

      // => ADD MORE HERE <=
    }
  })

  return Object.values(employeeMapping).map(employeeData => {
    const employeeDict = employeeData.employee.toJSON()
    return Object.assign(employeeDict, _.omit(employeeData, ['employee']))
  })

  // TODO => Get data between 26-25 each month

}
