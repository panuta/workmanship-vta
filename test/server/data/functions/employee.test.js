import moment from 'moment'

import Config from '../../../../server/config'
import { initDatabase, dropDatabase, Employee } from '../../../../server/data/models'
import { findEmployees } from '../../../../server/data/functions/employee'

jest.mock('../../../../server/config')

describe('Employee functions', () => {
  beforeAll(async () => {
    Config.__mockConfig({
      databaseStorage: ':memory:',
      databaseTimestamps: false
    })
    await initDatabase()
  })

  afterAll(async () => {
    await dropDatabase()
  })

  beforeEach(async () => {
    jest.resetAllMocks()
    await Employee.destroy({ where: {} })
  })

  const initEmployeesTables = async () => {
    const employeesData = [{
      code: '0001', fullName: 'Employee1',
      company: 'SE', status: '', department: '',
      startDate: new Date(2020, 1, 1),
      terminationDate: new Date(2020, 3, 30)
    }, {
      code: '0002', fullName: 'Employee2',
      company: 'SE', status: '', department: '',
      startDate: new Date(2020, 1, 1),
      terminationDate: new Date(2020, 8, 30)
    }, {
      code: '0003', fullName: 'Employee3',
      company: 'SE', status: '', department: '',
      startDate: new Date(2020, 0, 1),
      terminationDate: null
    }, {
      code: '0004', fullName: 'Employee4',
      company: 'SE', status: '', department: '',
      startDate: new Date(2020, 9, 1),
      terminationDate: null
    }]
    await Employee.bulkCreate(employeesData)
  }

  describe('findEmployees function', () => {
    it('should return all employees', async () => {
      await initEmployeesTables()
      await expect(findEmployees()).resolves.toMatchSnapshot()
    })

    it('should return active employees', async () => {
      await initEmployeesTables()
      await expect(findEmployees(moment({ year: 2020, month: 8 }))).resolves.toMatchSnapshot()
    })
  })
})
