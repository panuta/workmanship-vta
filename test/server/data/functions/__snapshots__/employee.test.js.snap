// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employee functions findEmployees function should return active employees 1`] = `
Array [
  Object {
    "code": "0002",
    "company": "SE",
    "department": "",
    "fullName": "Employee2",
    "id": 6,
    "nickName": null,
    "position": null,
    "salutation": null,
    "startDate": "2020-02-01",
    "status": "",
    "terminationDate": "2020-09-30",
  },
  Object {
    "code": "0003",
    "company": "SE",
    "department": "",
    "fullName": "Employee3",
    "id": 7,
    "nickName": null,
    "position": null,
    "salutation": null,
    "startDate": "2020-01-01",
    "status": "",
    "terminationDate": null,
  },
]
`;

exports[`Employee functions findEmployees function should return all employees 1`] = `
Array [
  Object {
    "code": "0001",
    "company": "SE",
    "department": "",
    "fullName": "Employee1",
    "id": 1,
    "nickName": null,
    "position": null,
    "salutation": null,
    "startDate": "2020-02-01",
    "status": "",
    "terminationDate": "2020-04-30",
  },
  Object {
    "code": "0002",
    "company": "SE",
    "department": "",
    "fullName": "Employee2",
    "id": 2,
    "nickName": null,
    "position": null,
    "salutation": null,
    "startDate": "2020-02-01",
    "status": "",
    "terminationDate": "2020-09-30",
  },
  Object {
    "code": "0003",
    "company": "SE",
    "department": "",
    "fullName": "Employee3",
    "id": 3,
    "nickName": null,
    "position": null,
    "salutation": null,
    "startDate": "2020-01-01",
    "status": "",
    "terminationDate": null,
  },
  Object {
    "code": "0004",
    "company": "SE",
    "department": "",
    "fullName": "Employee4",
    "id": 4,
    "nickName": null,
    "position": null,
    "salutation": null,
    "startDate": "2020-10-01",
    "status": "",
    "terminationDate": null,
  },
]
`;
